pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/api/core/v1"
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	informers "k8s.io/client-go/informers/core/v1"
+	networkinginformers "k8s.io/client-go/informers/networking/v1alpha1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	nodeutil "k8s.io/component-helpers/node/util"
+	"k8s.io/klog/v2"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+	controllerutil "k8s.io/kubernetes/pkg/controller/util/node"
+	"k8s.io/kubernetes/pkg/util/slice"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultClusterCIDRKey        = "kubernetes.io/clusterCIDR"
+	defaultClusterCIDRValue      = "default"
+	defaultClusterCIDRName       = "default-cluster-cidr"
+	defaultClusterCIDRAPIVersion = "networking.k8s.io/v1alpha1"
+	clusterCIDRFinalizer         = "networking.k8s.io/cluster-cidr-finalizer"
+	ipv4MaxCIDRMask              = 32
+	ipv6MaxCIDRMask              = 128
+	minPerNodeHostBits           = 4
+	maxRetries                   = 15
+)
+
+// CIDRs are reserved, then node resource is patched with them.
+// multiCIDRNodeReservedCIDRs holds the reservation info for a node.
+type multiCIDRNodeReservedCIDRs struct {
+	nodeReservedCIDRs
+	clusterCIDR *cidrset.ClusterCIDR
+}
+
+type multiCIDRRangeAllocator struct {
+	client clientset.Interface
+	// nodeLister is able to list/get nodes and is populated by the shared informer passed to controller.
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	nodesSynced cache.InformerSynced
+	// clusterCIDRLister is able to list/get clustercidrs and is populated by the shared informer passed to controller.
+	clusterCIDRLister networkinglisters.ClusterCIDRLister
+	// clusterCIDRSynced returns true if the clustercidr shared informer has been synced at least once.
+	clusterCIDRSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
```

Comment:
heh, now I understand it, this is coming from

https://github.com/kubernetes/kubernetes/blob/bd9444c1cfd438690dc21c31404dd79ab60dc6cb/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L72-L76

Reply:
And here: https://github.com/kubernetes/kubernetes/blob/137439194d8e27f34081218a7e8cfb1e31e74a69/pkg/controller/nodeipam/ipam/range_allocator.go#L61-L63

Based the controller on these existing range allocators and avoided re writing the controllers, given my limited knowledge about that, being used kubebuilder before to auto generate them :D

Reply:
No worries, that happens to me every now and then too